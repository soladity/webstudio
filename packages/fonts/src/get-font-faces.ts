import { FONT_FORMATS } from "./constants";
import type { FontMeta, FontFormat, FontMetaStatic } from "./schema";

export type PartialFontAsset = {
  format: FontFormat;
  meta: FontMeta;
  path: string;
};

export type FontFace = {
  fontFamily: string;
  fontDisplay: "swap" | "auto" | "block" | "fallback" | "optional";
  src: string;
  fontStyle?: FontMetaStatic["style"];
  fontWeight?: number | string;
  fontStretch?: string;
};

const formatFace = (asset: PartialFontAsset, format: string): FontFace => {
  if ("variationAxes" in asset.meta) {
    const { wght, wdth } = asset.meta?.variationAxes ?? {};
    return {
      fontFamily: asset.meta.family,
      fontStyle: "normal",
      fontDisplay: "swap",
      src: `url('${asset.path}') format('${format}')`,
      fontStretch: wdth ? `${wdth.min}% ${wdth.max}%` : undefined,
      fontWeight: wght ? `${wght.min} ${wght.max}` : undefined,
    };
  }
  return {
    fontFamily: asset.meta.family,
    fontStyle: asset.meta.style,
    fontWeight: asset.meta.weight,
    fontDisplay: "swap",
    src: `url('${asset.path}') format('${format}')`,
  };
};

const getKey = (asset: PartialFontAsset) => {
  if ("variationAxes" in asset.meta) {
    return asset.meta.family + Object.values(asset.meta.variationAxes).join("");
  }
  return asset.meta.family + asset.meta.style + asset.meta.weight;
};

export const getFontFaces = (
  assets: Array<PartialFontAsset>
): Array<FontFace> => {
  const faces = new Map();
  for (const asset of assets) {
    const assetKey = getKey(asset);
    const face = faces.get(assetKey);
    const format = FONT_FORMATS.get(asset.format);
    if (format === undefined) {
      // Should never happen since we allow only uploading formats we support
      continue;
    }

    if (face === undefined) {
      const face = formatFace(asset, format);
      faces.set(assetKey, face);
      continue;
    }

    // We already have that font face, so we need to add the new src
    face.src += `, url('${asset.path}') format('${format}')`;
  }
  return Array.from(faces.values());
};
