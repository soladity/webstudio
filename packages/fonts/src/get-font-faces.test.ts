import { describe, test, expect } from "@jest/globals";
import { getFontFaces, type PartialFontAsset } from "./get-font-faces";

describe("getFontFaces()", () => {
  test("different formats", () => {
    const assets: Array<PartialFontAsset> = [
      {
        format: "woff",
        meta: {
          family: "Roboto",
          style: "normal",
          weight: 400,
        },
        name: "roboto.woff",
      },
      {
        format: "ttf",
        meta: {
          family: "Roboto",
          style: "normal",
          weight: 400,
        },
        name: "roboto.ttf",
      },
    ];
    expect(getFontFaces(assets, { assetBaseUrl: "/fonts/" })).toMatchSnapshot();
  });

  test("different style", () => {
    const assets: Array<PartialFontAsset> = [
      {
        format: "ttf",
        meta: {
          family: "Roboto",
          style: "normal",
          weight: 400,
        },
        name: "roboto.ttf",
      },
      {
        format: "ttf",
        meta: {
          family: "Roboto",
          style: "italic",
          weight: 400,
        },
        name: "roboto-italic.ttf",
      },
    ];
    expect(getFontFaces(assets, { assetBaseUrl: "/fonts/" })).toMatchSnapshot();
  });

  test("different weight", () => {
    const assets: Array<PartialFontAsset> = [
      {
        format: "ttf",
        meta: {
          family: "Roboto",
          style: "normal",
          weight: 400,
        },
        name: "roboto.ttf",
      },
      {
        format: "ttf",
        meta: {
          family: "Roboto",
          style: "normal",
          weight: 500,
        },
        name: "roboto-bold.ttf",
      },
    ];
    expect(getFontFaces(assets, { assetBaseUrl: "/fonts/" })).toMatchSnapshot();
  });

  test("variable font", () => {
    const assets: Array<PartialFontAsset> = [
      {
        format: "ttf",
        meta: {
          family: "Inter",
          variationAxes: {
            wght: { name: "wght", min: 100, default: 400, max: 1000 },
            wdth: { name: "wdth", min: 25, default: 100, max: 151 },
            opsz: { name: "opsz", min: 8, default: 14, max: 144 },
            GRAD: { name: "GRAD", min: -200, default: 0, max: 150 },
            slnt: { name: "slnt", min: -10, default: 0, max: 0 },
            XTRA: { name: "XTRA", min: 323, default: 468, max: 603 },
            XOPQ: { name: "XOPQ", min: 27, default: 96, max: 175 },
            YOPQ: { name: "YOPQ", min: 25, default: 79, max: 135 },
            YTLC: { name: "YTLC", min: 416, default: 514, max: 570 },
            YTUC: { name: "YTUC", min: 528, default: 712, max: 760 },
            YTAS: { name: "YTAS", min: 649, default: 750, max: 854 },
            YTDE: { name: "YTDE", min: -305, default: -203, max: -98 },
            YTFI: { name: "YTFI", min: 560, default: 738, max: 788 },
          },
        },
        name: "inter.ttf",
      },
    ];
    expect(getFontFaces(assets, { assetBaseUrl: "/fonts/" })).toMatchSnapshot();
  });
});
