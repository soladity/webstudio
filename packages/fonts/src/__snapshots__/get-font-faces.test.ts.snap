// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFontFaces() different formats 1`] = `
[
  {
    "fontDisplay": "swap",
    "fontFamily": "Roboto",
    "fontStyle": "normal",
    "fontWeight": 400,
    "src": "url('/fonts/roboto.woff') format('woff'), url('/fonts/roboto.ttf') format('truetype')",
  },
]
`;

exports[`getFontFaces() different style 1`] = `
[
  {
    "fontDisplay": "swap",
    "fontFamily": "Roboto",
    "fontStyle": "normal",
    "fontWeight": 400,
    "src": "url('/fonts/roboto.ttf') format('truetype')",
  },
  {
    "fontDisplay": "swap",
    "fontFamily": "Roboto",
    "fontStyle": "italic",
    "fontWeight": 400,
    "src": "url('/fonts/roboto-italic.ttf') format('truetype')",
  },
]
`;

exports[`getFontFaces() different weight 1`] = `
[
  {
    "fontDisplay": "swap",
    "fontFamily": "Roboto",
    "fontStyle": "normal",
    "fontWeight": 400,
    "src": "url('/fonts/roboto.ttf') format('truetype')",
  },
  {
    "fontDisplay": "swap",
    "fontFamily": "Roboto",
    "fontStyle": "normal",
    "fontWeight": 500,
    "src": "url('/fonts/roboto-bold.ttf') format('truetype')",
  },
]
`;

exports[`getFontFaces() variable font 1`] = `
[
  {
    "fontDisplay": "swap",
    "fontFamily": "Inter",
    "fontStretch": "25% 151%",
    "fontStyle": "normal",
    "fontWeight": "100 1000",
    "src": "url('/fonts/inter.ttf') format('truetype')",
  },
]
`;
